### PR TITLE
Share more constant initializers

### PR DESCRIPTION
**Pull request by pengwa**
### Share more constant initializers.

`ConstantSharing` transformer originally only handle single value initializer (scalar or 1D). 

This PR tried to share more cases to make common subexpression elimination transformer to remove more duplicated nodes. 

Originally, we used a single vector<std::variant<float,half,int32,int64>> to store different scalar values. In this PR, we create a unordered map with its key being data_type + rank + element count, and its value is a vector of `InitializerValue`.

For one specific initializer, if it fulfils the condition, then finally will find the corresponding vector of `InitializerValue` by its <data_type + rank + element count>, then search from the vector whether the constant tensor already exist or not. After that, a value id is returned, which will be combined together with <data_type + rank + element count> to form the pattern key to decide which tensor to reuse (legacy code).

### Motivation and Context

One example we see here is:

```mermaid
stateDiagram
    [*] --> LayerNorm(b,s,64)
    LayerNorm(b,s,64) --> Reshape1
    Shape1_Const[b*s,64] --> Reshape1

    LayerNorm(b,s,64) --> Reshape2
    Shape2_Const[b*s,64] --> Reshape2


    Reshape1 --> AttentionSubGraph
    Reshape2 -->  Add
    AttentionSubGraph--> Add
   Add --> [*]
```

Ideally CommonSubexpressionElimination can remove one of `Reshape1` and `Reshape2`, while since `Shape1_Const` and `Shape2_Const` are different NodeArg*, so it did not remove the duplication. 

This is an example: removing the duplication will bring more opportunities to apply graph transformations.

### CODE REVIEW

File: onnxruntime/core/optimizer/constant_sharing.cc [hunk 1]
```diff
@@ -29,20 +29,35 @@ bool IsSupportedDataType(int32_t data_type) {
 
 using SupportedTypeList = boost::mp11::mp_list<MLFloat16, float, double, int32_t, int64_t>;
 
-bool IsValidSingleValueShape(const ONNX_NAMESPACE::TensorShapeProto* input_shape) {
+static constexpr int64_t MAX_SIZE_PER_VALUE = 8;
```
Comment (askhade) [inline onnxruntime/core/optimizer/constant_sharing.cc:32]: why do we have this restriction on num of elements?

Comment (pengwa) [reply onnxruntime/core/optimizer/constant_sharing.cc:32]: Having a bigger tensor element size threshold here means more overhead running ConstantSharing graph transformation. Originally the number is 1, now I changed it to 8 gradually. Maybe we can make it bigger once we found it helps for some specific scenarios.

Comment (askhade) [reply onnxruntime/core/optimizer/constant_sharing.cc:32]: Can you add some cooments around the reasoning behind choosing 8 and what should one consider if they want to change this or remove this limitation altogether in future. Thanks!

Comment (pengwa) [reply onnxruntime/core/optimizer/constant_sharing.cc]: Sure, I added a comment for it. :)

**Pull request merged**